jd4/daemon.py

```diff
@@ -39,20 +42,23 @@ def __init__(self, session, request, ws):
                 else:
                     raise SystemError('Unknown event: {}'.format(event))
             else:
-                type = self.request.pop('type')
-                if type == 0:
+                self.type = self.request.pop('type')
```

Comment:
move this into `__init__`.

also note that the `problem_data_change` event should not be handled by `JudgeHandler`.